NonBayesian/Second/second2.py

```diff
@@ -0,0 +1,226 @@
+                for k in range(1, k + 1):
+                    cache_arr[i, j, k] = cache_arr[i-1, j ,k] + cache_arr[i, j-1, k] + cache_arr[i, j, k-1] \
+                                         - cache_arr[i, j-1, k-1] - cache_arr[i-1, j, k-1] - cache_arr[i-1, j-1, k]\
+                                         + cache_arr[i-1, j-1, k-1] + arr[i-1, j-1, k-1]
+        return cache_arr
+        return 0
+    else:
+        raise Exception("Bad dimension")
+
+
+def cumsum_cached(arr, cache_arr, start_point, end_point):
+    """
+    >>> cumsum_cached(np.ones(10), cache(np.ones(10)), [7], [3])
+    -3.0
+    >>> cumsum_cached(np.zeros(10), cache(np.zeros(10)), [10], [0])
+    0.0
+    """
+    if arr.ndim == 1:
+        return cache_arr[end_point[0]] - cache_arr[start_point[0]-1]
+
+    if arr.ndim == 2:
+        start_point = start_point - np.array([1, 1])
+        return cache_arr[ end_point[0], end_point[1] ]\
+               - cache_arr[ end_point[0], start_point[1] ]\
+               - cache_arr[ start_point[0], end_point[1] ]\
+               + cache_arr[ start_point[0], start_point[1] ]
+
+    if arr.ndim == 3:
+        start_point = start_point - np.array([1, 1, 1])
+        return cache_arr[end_point[0], end_point[1], end_point[2]] \
+               - cache_arr[start_point[0], end_point[1], end_point[2]] \
+               - cache_arr[end_point[0], start_point[1], end_point[2]] \
+               - cache_arr[end_point[0], end_point[1], start_point[2]] \
+               + cache_arr[end_point[0], start_point[1], start_point[2]] \
+               + cache_arr[start_point[0], end_point[1], start_point[2]] \
+               + cache_arr[start_point[0], start_point[1], end_point[2]] \
+               - cache_arr[start_point[0], start_point[1], start_point[2]]
+    else:
+        raise Exception("Bad dimension")
+
+
+def cumsum_slow(arr, start_point, end_point):
+    """
+        >>> cumsum_slow(np.ones(10), [3], [7])
+        5.0
+        >>> cumsum_slow(np.zeros(10), [10], [0])
+        0
+    """
+    result = 0
+
+    if arr.ndim == 1:
+        for i in range(start_point[0], end_point[0]+1):
+            result = result + arr[i]
+        return result
+
+    if arr.ndim == 2:
+        for i in range(start_point[0], end_point[0]+1):
+            for j in range(start_point[1], end_point[1]+1):
+                result = result + arr[i, j]
+        return result
+
+    if arr.ndim == 3:
+        for i in range(start_point[0], end_point[0]+1):
+            for j in range(start_point[1], end_point[1]+1):
+                for k in range(start_point[2], end_point[2]+1):
+                    result = result + arr[i, j, k]
+        return result
+    else:
+        raise Exception("Bad dimension")
+
+def experiment(t, cache_arr):
+    start_point = [0, 0]
+    end_point = [t - 1, t - 1]
+
+    start = timer()
+    cumsum_cached(A, cache_arr, start_point, end_point)
+    end = timer()
+    t1 = end - start
+
+    return (t1*1000)
```

Comment:
Make this more readable and less redundant
```suggestion
    return t1*1E3
```